parlai/core/worlds.py

```diff
@@ -1210,6 +1218,115 @@ def report(self):
         return self.world.report()
 
 
+class BackgroundDriverWorld(World):
+    def __init__(self, opt: Opt, world: World):
+        self.world = world
+        super().__init__(opt, agents=world.agents, shared=None)
+
+        import torch.multiprocessing as mp
+
+        self._num_workers = self.opt['num_workers']
+        self._process_queue = mp.Queue(maxsize=4 * self._num_workers)
+        self._process_pool = self._start_processes()
+
+        self._batch_buffer = []
+        self.metrics = TeacherMetrics()
+
+    def _start_processes(self):
+        import torch.multiprocessing as mp
+
+        return mp.start_processes(
+            fn=BackgroundWorkerDynamicBatchWorld.launch_process,
+            nprocs=self._num_workers,
+            args=(self.opt, self.get_model_agent(), self._process_queue),
+            join=False,
+            start_method='fork',
+        )
+
+    def reset(self):
+        """
+        Reset all subworlds.
+        """
+        self.world.reset()
+
+    def reset_metrics(self):
+        """
+        Reset metrics in all subworlds.
+        """
+        self.world.reset_metrics()
+        self.metrics.clear()
+
+    def get_task_agent(self):
+        return self.world.get_task_agent()
+
+    def get_model_agent(self):
+        return self.world.get_model_agent()
+
+    def num_examples(self):
+        return self.world.num_examples()
+
+    def num_episodes(self):
+        return self.world.num_episodes()
+
+    def parley(self):
+        index, batch = self._process_queue.get()
+        response_object = self.get_model_agent().batch_act(batch)
+        # compute metrics
+        for response in response_object:
+            self.metrics.evaluate_response(response, [])
+        self.total_parleys += 1
+        self.total_exs += batch.batchsize
+
+    def get_total_exs(self):
+        return self.total_exs
+
+    def get_total_epochs(self):
+        return self.total_exs / self.num_examples()
+
+    def report(self):
+        return aggregate_unnamed_reports([self.world.report(), self.metrics.report()])
+
+    def __del__(self):
+        logging.debug("Killing all the worker processes")
+        for p in self._process_pool.processes:
+            p.kill()
+
+
+class BackgroundWorkerDynamicBatchWorld(DynamicBatchWorld):
+    @classmethod
+    def launch_process(cls, index, opt, model_agent, process_queue):
```

Comment:
oh is index the implicit first arg for `mp.start_processes`?

Reply:
yes

Reply:
added a comment